File: universal-utils/src/main/scala/io/github/ppdzm/utils/universal/feature/LoanPattern.scala
package io.github.ppdzm.utils.universal.feature

import scala.concurrent.{ExecutionContext, Future}
import scala.language.reflectiveCalls
import scala.util.control.Exception._

/**
 * Created by Stuart Alex on 2017/3/20.
 *
 * Loan pattern implementation
 */
object LoanPattern {
    type Closable = {def close()}

    def using[R <: Closable, A](resource: R)(f: R => A): A = {
        try {
            f(resource)
        } finally {
            ignoring(classOf[Throwable]) apply {
                resource.close()
            }
        }
    }

    /**
     * Guarantees a Closeable resource will be closed after being passed to a block that takes
     * the resource as a parameter and returns a Future.
     */
    def futureUsing[R <: Closable, A](resource: R)(f: R => Future[A])(implicit ec: ExecutionContext): Future[A] = {
        f(resource) andThen { case _ => resource.close() } // close no matter what
    }

}

File: universal-utils/src/main/java/io/github/ppdzm/utils/universal/feature/LoanPattern.java
package io.github.ppdzm.utils.universal.feature;

import java.util.function.Function;

/**
 * @author Created by Stuart Alex on 2023/5/13.
 */
public class LoanPattern {

    public static <T extends AutoCloseable, R> R using(T resource, Function<T, R> block) throws Exception {
        try {
            return block.apply(resource);
        } finally {
            if (resource != null) {
                resource.close();
            }
        }
    }

}
